### PR TITLE
Added clearSession iOS Safari Method

### PR DESCRIPTION
**Pull request by cocojoe**
Added clearSession JS wrapper

### CODE REVIEW

**Comment by cocojoe** [comment]
Also ran it through Xcode/Re-Indent 😄 
So wondering should this be a `Promise` vs a `Callback`?
I didn't make the SFView semi-transparent as the more I use it, I feel by showing the browser it shows more clearly to the user that it's doing something.

**Comment by cocojoe** [comment]
Note: Sample PR https://github.com/auth0-samples/auth0-react-native-sample/pull/4

File: webauth/index.js [hunk 1]
```diff
@@ -96,4 +96,21 @@ export default class WebAuth {
...
+   * @param {Bool} [parameters.federated] Optionally remove the IdP session.
+   * @returns {Callback}
+   * @see https://auth0.com/docs/logout
+   *
+   * @memberof WebAuth
+   */
+  clearSession(options = {}, callback) {
+    const { clientId, domain, client, agent } = this;
+    const federated = options.federated || false;
+    A0Auth0.clearSession(domain, federated, callback);
```
Comment (hzalaz) [inline webauth/index.js]: Since this method is not yet available on android lets fail when the OS is not iOS.

File: webauth/index.js [hunk 2]
```diff
@@ -96,4 +96,21 @@ export default class WebAuth {
...
+   *  In iOS it will use `SFSafariViewController`
+   *
+   * @param {Object} parameters parameters to send
+   * @param {Bool} [parameters.federated] Optionally remove the IdP session.
+   * @returns {Callback}
+   * @see https://auth0.com/docs/logout
+   *
+   * @memberof WebAuth
+   */
+  clearSession(options = {}, callback) {
```
Comment (hzalaz) [inline webauth/index.js]: The library uses promises instead of callbacks

File: webauth/index.js (hunk 1 above)
Comment (hzalaz) [inline webauth/index.js]: Some help https://facebook.github.io/react-native/docs/platform-specific-code.html

**Comment by cocojoe** [comment]
Ping

File: ios/A0Auth0.m [hunk 3]
```diff
@@ -29,35 +30,49 @@ - (dispatch_queue_t)methodQueue
...
-    controller.delegate = self;
-    [self terminateWithError:RCTMakeError(@"Only one Safari can be visible", nil, nil) dismissing:YES animated:NO];
-    [window.rootViewController presentViewController:controller animated:YES completion:nil];
-    self.last = controller;
+    [self presentSafariWithURL:[NSURL URLWithString:urlString]];
     self.sessionCallback = callback;
 }
 
+RCT_EXPORT_METHOD(clearSession:(NSString *)domain federated:(BOOL)federated callback:(RCTResponseSenderBlock)callback) {
+    NSString *urlString = [NSString stringWithFormat:@"https://%@/v2/logout", domain];
```
Comment (hzalaz) [inline ios/A0Auth0.m]: Lets use NSURLComponents to build urls

Comment (hzalaz) [inline ios/A0Auth0.m]: And TBH I'd rather build the full url in JS  and make this method a more generic one like show and close on load thing

File: ios/A0Auth0.m [hunk 4]
```diff
@@ -101,31 +116,33 @@ - (NSString *)sign:(NSString*)value {
...
+                            @"error": @"a0.session.user_cancelled",
+                            @"error_description": @"User cancelled the Auth"
+                            };
     [self terminateWithError:error dismissing:NO animated:NO];
 }
 
 - (void)safariViewController:(SFSafariViewController *)controller didCompleteInitialLoad:(BOOL)didLoadSuccessfully {
-    if (!didLoadSuccessfully) {
+    if (self.didLoadCallback) {
+        didLoadSuccessfully ? self.didLoadCallback(@[[NSNull null]]) : self.didLoadCallback(@[@{@"error": @"failed to load"}]);
```
Comment (hzalaz) [inline ios/A0Auth0.m]: Probably cleaner to do the ternary operator with the parameter and the call the callback later

File: ios/A0Auth0.m [hunk 5]
```diff
@@ -101,31 +116,33 @@ - (NSString *)sign:(NSString*)value {
...
+        self.didLoadCallback = nil;
+    } else if (!didLoadSuccessfully) {
         NSDictionary *error = @{
-            @"error": @"a0.session.failed_load",
-            @"error_description": @"Failed to load authorize url"
-        };
+                                @"error": @"a0.session.failed_load",
+                                @"error_description": @"Failed to load authorize url"
+                                };
         [self terminateWithError:error dismissing:YES animated:YES];
```
Comment (hzalaz) [inline ios/A0Auth0.m]: We need to clean the didLoadCallback too

File: webauth/index.js [hunk 6]
```diff
@@ -96,4 +96,27 @@ export default class WebAuth {
...
+   *
+   * @memberof WebAuth
+   */
+  clearSession(options = {}) {
+    const { clientId, domain, client, agent } = this;
+    const federated = options.federated || false;
+    return new Promise(function (resolve, reject) {
+      if (Platform.OS !== 'ios') { reject(new Error('clearSession only supported in iOS')); }
+      A0Auth0.clearSession(domain, federated, function (err, data) {
+        if (err !== null) return reject(err);
```
Comment (hzalaz) [inline webauth/index.js]: Always use `{}`

File: ios/A0Auth0.m (hunk 5 above)
Comment (cocojoe) [inline ios/A0Auth0.m]: What are you thinking here, refactor it all to use the `didLoadCallback` and make the auto close optional?

Comment (cocojoe) [inline ios/A0Auth0.m]: Ignore that, out of scope of this PR.

**Pull request merged**